### PR TITLE
[bundlephobia] Added a shield for calculating npm bundle sizes

### PR DESCRIPTION
**Pull request by pastelsky**
A shield for bundle sizes of front end npm packages, as discussed over here - https://github.com/badges/shields/issues/941

### CODE REVIEW

**Comment by pastelsky** [comment]
I'm not exactly sure why the CI check failed. Looks like some sort of an error related to `svg2img` library, which isn't related to this PR.

**Comment by paulmelnikow** [comment]
Yea, you're right, it's unrelated. I opened #1392 to increase the timeout on that test. It times out fairly often.

**Comment by pastelsky** [comment]
Hey, are there any blockers for this to get merged?

File: lib/all-badge-examples.js [hunk 1]
```diff
@@ -781,6 +781,26 @@ const allBadgeExamples = [
...
+        ]
+      },
+      {
+        title: 'npm bundle size (minified + gzip)',
+        previewUri: '/bundlephobia/gzip/react.svg',
+        keywords: [
+          'node',
+          'npm',
+          'bundle',
+          'size'
```
Comment (paulmelnikow) [inline lib/all-badge-examples.js]: You can just include `'node'` here and remove the others. Substrings of the title will be matched automatically.

File: server.js [hunk 2]
```diff
@@ -7715,6 +7715,53 @@ camp.route(/^\/nsp\/npm\/(?:@([^/]+)?\/)?([^/]+)?(?:\/([^/]+)?)?\.(svg|png|gif|j
...
 
+// bundle size for npm packages
+camp.route(/^\/bundlephobia\/(min|gzip)\/(?:@([^/]+)?\/)?([^/]+)?(?:\/([^/]+)?)?\.(svg|png|gif|jpg|json)?$/,
+  cache((data, match, sendBadge, request) => {
+  // A: /bundlephobia/(min|gzip)/:package.:format
+  // B: /bundlephobia/(min|gzip)/:package/:version.:format
+  // C: /bundlephobia/(min|gzip)/@:scope/:package.:format
+  // D: /bundlephobia/(min|gzip)/@:scope/:package/:version.:format
+  const capturedResultType = match[1];
+  const capturedScopeWithoutAtSign = match[2];
```
Comment (paulmelnikow) [inline server.js]: Much as I love verbose names, I think you could drop `captured` from these, and shorten this one to `scope`.

File: server.js [hunk 3]
```diff
@@ -7715,6 +7715,53 @@ camp.route(/^\/nsp\/npm\/(?:@([^/]+)?\/)?([^/]+)?(?:\/([^/]+)?)?\.(svg|png|gif|j
   }
 }));
 
+// bundle size for npm packages
+camp.route(/^\/bundlephobia\/(min|gzip)\/(?:@([^/]+)?\/)?([^/]+)?(?:\/([^/]+)?)?\.(svg|png|gif|jpg|json)?$/,
```
Comment (paulmelnikow) [inline server.js]: Minified and gzipped aren't exclusive. Does gzip mean minified and gzipped?

File: server.js [hunk 4]
```diff
@@ -7715,6 +7715,53 @@ camp.route(/^\/nsp\/npm\/(?:@([^/]+)?\/)?([^/]+)?(?:\/([^/]+)?)?\.(svg|png|gif|j
...
+        badgeData.colorscheme = 'red';
+      } else {
+        badgeData.text[1] = prettyBytes(showMin ? body.size : body.gzip);
+        badgeData.colorscheme = 'blue';
+      }
+      sendBadge(capturedFormat, badgeData);
+    });
+  }
+
+  getBundlephobiaResults(capturedScopeWithoutAtSign, capturedPackageName, capturedVersion);
```
Comment (paulmelnikow) [inline server.js]: What advantage does making this basically an IIFE provide?

File: server.js [hunk 5]
```diff
@@ -7715,6 +7715,53 @@ camp.route(/^\/nsp\/npm\/(?:@([^/]+)?\/)?([^/]+)?(?:\/([^/]+)?)?\.(svg|png|gif|j
...
+      json: true,
+    };
+
+    request(requestOptions, (error, response, body) => {
+      if (error !== null || typeof body !== 'object' || body === null || body.error) {
+        badgeData.text[1] = body.error.code || 'error';
+        badgeData.colorscheme = 'red';
+      } else {
+        badgeData.text[1] = prettyBytes(showMin ? body.size : body.gzip);
+        badgeData.colorscheme = 'blue';
```
Comment (paulmelnikow) [inline server.js]: Could you put this in a try block to handle the case where the JSON is malformed?

File: service-tests/bundlephobia.js [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+      .get(withoutErrorGzip)
+      .expectJSONTypes(Joi.object().keys({ name: 'gzip size', value: withoutErrorSizeGzip }))
+  }
+
+  if (typeof noExistPackage === 'string') {
+    t.create(`Format '${formats[format]}' where it doesn't exist`)
+      .get(noExistPackage)
+      .expectJSON({
+        name: 'minified size',
+        value: 'PackageNotFoundError',
```
Comment (paulmelnikow) [inline service-tests/bundlephobia.js]: **minified size | package not found** or even **bundlephobia | package not found** would be more stylistically consistent with our error badges.

File: service-tests/bundlephobia.js [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+  A: Joi.string().regex(/^\d+[.]?\d+ kB$/),
+  B: '8.94 kB',
+  C: Joi.string().regex(/^\d+[.]?\d+ kB$/),
+  D: '3.51 kB',
+}
+
+const withoutErrorSizesGzip = {
+  A: Joi.string().regex(/^\d+[.]?\d+ kB$/),
+  B: '3.58 kB',
+  C: Joi.string().regex(/^\d+[.]?\d+ kB$/),
```
Comment (paulmelnikow) [inline service-tests/bundlephobia.js]: You could use [`isFileSize`](https://github.com/badges/shields/blob/master/service-tests/helpers/validators.js#L53) for this.

File: service-tests/bundlephobia.js [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+  B: '/min/some-no-exist/1.0.0.json',
+  C: '/min/@some-no-exist/some-no-exist.json',
+}
+
+Object.keys(formats).forEach(format => {
+  const withoutErrorMin = withoutErrorsMin[format]
+  const withoutErrorSizeMin = withoutErrorSizesMin[format]
+
+  const withoutErrorGzip = withoutErrorsGzip[format]
+  const withoutErrorSizeGzip = withoutErrorSizesGzip[format]
```
Comment (paulmelnikow) [inline service-tests/bundlephobia.js]: This is nice and DRY, though it's a bit difficult to follow what you're doing. Also not sure it's necessary to be exhaustive in testing these combinations.

Could I suggest a couple ways of making it easier to understand? One would be to reduce this to 5 success cases (four min and one gzip) and 1–3 failure cases, and lay out each test explicitly. The other would be to basically invert your data objects, making an array of objects with `format`, `withoutErrorMin`, etc. as keys.

File: server.js (hunk 3 above)
Comment (pastelsky) [reply server.js]: Yes, it does.

File: server.js (hunk 2 above)
Comment (pastelsky) [reply server.js]: Cool.

File: server.js (hunk 4 above)
Comment (pastelsky) [reply server.js]: Done.

File: lib/all-badge-examples.js (hunk 1 above)
Comment (pastelsky) [reply lib/all-badge-examples.js]: Done.

File: server.js (hunk 5 above)
Comment (pastelsky) [reply server.js]: Added a check for `typeof body !== 'object' || body === null`

File: service-tests/bundlephobia.js (hunk 6 above)
Comment (pastelsky) [reply service-tests/bundlephobia.js]: These error codes are returned by the bundlephobia service, and there are multiple such types - `EntryPointError` | `PackageNotFoundError` | `VersionMismatchError` etc.  I'd prefer to keep them the same to enable easier debugging when issues are filed.

File: server.js (hunk 3 above)
Comment (paulmelnikow) [reply server.js]: Then how about we make it `min` and `min-gzip`?

File: service-tests/bundlephobia.js (hunk 6 above)
Comment (paulmelnikow) [reply service-tests/bundlephobia.js]: I think it's sufficient to re-run the failing request when an issue is filed. One of the goals of Shields is to provide consistent and concise badges, which applies to error messages too.

**Comment by paulmelnikow** [comment]
Are you willing to make the change to standardize the errors? If not I could pick this up from here.

File: server.js (hunk 3 above)
Comment (pastelsky) [reply server.js]: Another commonly used term for this is "minzip"

Comment (paulmelnikow) [reply server.js]: Sure, that works and seems clear.

**Comment by pastelsky** [comment]
Since I might have new error types in the future, I've added a regex based formatter that basically does this conversion => `BuildError` to `build error`.  Unfortunately, I could not make time to refactor the tests. You can pick this up if you wish to, or I'll try to refactor them later when I find some time.

**Comment by paulmelnikow** [comment]
Okay, I'll pick that up.

I noticed the services-pr failure, which I think I fixed in #1454. If you merge master I think that should clear.

The failure main should be fixable with `npm run lint -- --fix`.

File: server.js (hunk 3 above)
Comment (pastelsky) [reply server.js]: Done.

**Comment by pastelsky** [comment]
Lint errors  / tests are passing now.

**Comment by tlaziuk** [comment]
@pastelsky @paulmelnikow - I want to help you with this PR, can I do that somehow?

**Comment by paulmelnikow** [comment]
Let's finish this up in #1481.

**Pull request closed without merging**